AMen to offload NCCL

CodersAcademy006 opened the pull request:
This pull request adds optional support for the AMem NCCL plugin to improve GPU memory efficiency during reinforcement learning training in Megatron-LM. By integrating `ncclPause()` and `ncclResume()`, the implementation enables offloading and reloading of NCCL-allocated memory, which is especially beneficial for large-scale distributed workloads. The integration is fully optional, controlled via command-line arguments, and maintains backward compatibility with existing training workflows.

The changes include a new AMem NCCL integration module, updates to initialization, RL utilities, and argument handling, as well as comprehensive documentation and an example training script to guide users. The implementation has been validated with error checks to ensure stability, making Megatron-LM more robust and memory-efficient for RL training scenarios.

Fix #2423

CodersAcademy006 committed:
Fix #2541: Make nv-grouped-gemm truly optional

- Move nv-grouped-gemm from dev/lts extras to new moe extra
- Users can now install megatron-core[dev] without build failures
- Add comprehensive error messages with installation instructions
- Update README with MoE dependencies documentation

This resolves the installation failure when nv-grouped-gemm cannot build
due to missing CUTLASS headers. Users who need MoE with grouped GEMM can
now explicitly install it with megatron-core[moe], while others can
install dev/lts extras without encountering build errors.

Fixes #2541

jon-barker commented:
Hi @CodersAcademy006 - sorry this didn't get attention sooner. I'd like to evaluate this approach for the native RL loop in megatron and shepherd it through the PR process if so. Thanks!

CodersAcademy006 commented:
Hi @jon-barker, thanks for reopening this! I'd be happy to update the PR to support the native RL loop.

Could you point me to the specific file or training step where the native RL loop resides? I can then move the `ncclPause/Resume` hooks to the appropriate location to ensure the memory is offloaded during generation.

CodersAcademy006 committed:
Resolve merge conflicts and ensure AMem NCCL integration is robust for RL workflows. Fixed imports, ensured native RL loop support, and validated no errors in touched files. Ready for review and CI on supported platforms.

CodersAcademy006 committed:
Merge branch 'main' into AMen-to-offloaf-NCCL

yobibyte commented on megatron/core/amem_nccl.py:
```diff
@@ -0,0 +1,236 @@
+        os.environ.setdefault('GMM_LOG', '3')
+
+        # Optionally set P2P buffer to be freed without CPU offload
+        # os.environ.setdefault('AMEM_NCCL_OFFLOAD_FREE_TAG', '7')
```
Remove/turn into an actionable TODO?

yobibyte commented on megatron/training/arguments.py:
```diff
@@ -2140,6 +2140,20 @@ def _add_rl_args(parser):
                         help='Skip BOS token at the beginning of the sequences. Default is False.')
+
+    # AMem NCCL plugin arguments
+    group.add_argument('--rl-use-amem-nccl', action='store_true',
```
Do we actually need two flags for use-amem and offload? Is there a valid scenario when we want to use --rl-use-amem-nccl flag, but do not want to offload?

yobibyte reviewed (commented):
Hey @CodersAcademy006, thanks for an update!

I did a review pass for the rl-related code, and left some comments. Please, have a look.

CodersAcademy006 committed:
Refactor AMem NCCL integration: remove redundant --rl-use-amem-nccl flag, gate all logic on --rl-amem-offload-during-rollout, encapsulate env setup, and update docs/examples per review

CodersAcademy006 committed:
Merge branch 'main' into AMen-to-offloaf-NCCL

CodersAcademy006 committed:
Resolve pyproject.toml merge conflict

- Keep AMem NCCL integration optional dependencies (moe and amem groups)
- Merge flash-linear-attention~=0.3.2 from origin/main into both dev and lts dependency groups
- Maintain all existing dependency structure

jon-barker committed:
Merge branch 'main' into AMen-to-offloaf-NCCL

jon-barker commented:
I hit an issue when attempting to test this functionality in megatron RL loop - AMEM doesn't currently support NCCL 2.28. I don't think it makes sense to merge this functionality until support for the latest NCCL is available.

CodersAcademy006 commented:
Following up ~6 months on: @jon-barker, has AMEM gained NCCL 2.28 support since your test? If this is unblocked now I'm happy to rebase (the branch has since gone stale/conflicting) and resolve conflicts so it can move forward. If it's still blocked on AMEM/NCCL, feel free to leave this parked and I'll rebase once support lands rather than churning it in the meantime.
